// @astrouxds/astro-web-components/dist/components/rux-tab-panels.js
import { attachShadow, createEvent, h, Host, proxyCustomElement } from '@stencil/core/internal/client';

const ruxTabPanelCss = ":host{display:block;position:relative;visibility:visible;z-index:1}:host([hidden]){display:none}:host(.hidden){display:none;position:absolute;visibility:hidden;z-index:-1}";

const RuxTabPanels$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    attachShadow(this);
    this.ruxRegisterPanels = createEvent(this, "ruxregisterpanels", 7);
  }
  connectedCallback() {
    this.el.setAttribute('style', 'position: relative; width: 100%;');
  }
  componentDidLoad() {
    this._registerTabPanels(this._getSlottedChildren());
  }
  _getSlottedChildren() {
    var _a, _b;
    const slot = (_b = (_a = this.el) === null || _a === void 0 ? void 0 : _a.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('slot');
    if (slot) {
      const childNodes = slot.assignedNodes({ flatten: true });
      const children = Array.prototype.filter.call(childNodes, (node) => node.nodeType == Node.ELEMENT_NODE);
      return children;
    }
    else {
      return [];
    }
  }
  _registerTabPanels(children) {
    this.ruxRegisterPanels.emit(children);
  }
  render() {
    return (h(Host, null, h("slot", null)));
  }
  get el() { return this; }
  static get style() { return ruxTabPanelCss; }
};

const RuxTabPanels = /*@__PURE__*/proxyCustomElement(RuxTabPanels$1, [1,"rux-tab-panels"]);

export { RuxTabPanels };
