// @astrouxds/astro-web-components/dist/components/rux-datetime.js
import { attachShadow, proxyCustomElement } from '@stencil/core/internal/client';

const RuxDatetime$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    attachShadow(this);
    /**
     * The date time to be formatted
     */
    this.date = new Date();
    /**
     * The locale
     */
    this.locale = 'default';
    /**
     * Display date in 12 hour time.
     */
    this.hour12 = false;
  }
  render() {
    const date = new Date(this.date);
    const { day, era, hour, hour12, locale, minute, month, second, timeZone, timeZoneName, weekday, year, } = this;
    return new Intl.DateTimeFormat(locale, {
      day,
      era,
      hour,
      hour12,
      minute,
      month,
      second,
      timeZone,
      timeZoneName,
      weekday,
      year,
    }).format(date);
  }
};

const RuxDatetime = /*@__PURE__*/proxyCustomElement(RuxDatetime$1, [1,"rux-datetime",{"date":[1],"locale":[1],"weekday":[1],"era":[1],"year":[1],"month":[1],"day":[1],"hour":[1],"minute":[1],"second":[1],"timeZoneName":[1,"time-zone-name"],"timeZone":[1,"time-zone"],"hour12":[4,"hour-12"]}]);

export { RuxDatetime };
