// @astrouxds/astro-web-components/dist/components/rux-tab.js
import { attachShadow, h, Host, proxyCustomElement } from '@stencil/core/internal/client';

const ruxTabCss = ":host{--tab-text-color:var(--color-primary);--tab-border-color:transparent;--tab-hover-text-color:var(--color-default-text);--tab-selected-text-color:var(--color-default-text);--tab-selected-border-color:var(--color-primary);box-sizing:border-box;display:flex;align-items:center;justify-content:center;padding:0 2rem;margin:0;min-width:5rem;text-decoration:none;color:var(--tab-text-color);border-bottom:5px solid var(--tab-border-color)}:host([hidden]){display:none}:host span{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}:host([small]){min-width:2rem;border-bottom:3px solid var(--tab-border-color)}:host([small][selected]){border-bottom:solid 3px var(--tab-selected-border-color)}:host([selected]){color:var(--tab-selected-text-color);border-bottom:5px solid var(--tab-selected-border-color)}:host(:hover){color:var(--tab-hover-text-color)}:host([disabled]){color:var(--tab-text-color);opacity:var(--disabled-opacity);cursor:not-allowed}";

const RuxTab$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    attachShadow(this);
    /**
     *  If present, overrides which tab is selected on load / mount. By default, the first <rux-tab> item is selected.
     */
    this.selected = false;
    /**
     * If present, sets a disabled state on this tab item, indicating it cannot be selected by user action.
     */
    this.disabled = false;
  }
  connectedCallback() {
    var _a, _b;
    this.el.setAttribute('role', 'tab');
    this.el.addEventListener('click', this._clickHandler);
    if ((_b = (_a = this.el) === null || _a === void 0 ? void 0 : _a.parentElement) === null || _b === void 0 ? void 0 : _b.getAttributeNode('small')) {
      this.el.setAttribute('small', '');
    }
  }
  _clickHandler(e) {
    if (this.disabled) {
      e.stopImmediatePropagation();
    }
  }
  render() {
    return (h(Host, { onClick: this._clickHandler }, h("slot", null)));
  }
  get el() { return this; }
  static get style() { return ruxTabCss; }
};

const RuxTab = /*@__PURE__*/proxyCustomElement(RuxTab$1, [1,"rux-tab",{"selected":[516],"disabled":[516]}]);

export { RuxTab };
