// @astrouxds/astro-web-components/dist/esm/rux-segmented-button.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-01dd0ed8.js';

const ruxSegmentedButtonCss = ":host{display:block;box-sizing:border-box;--segmented-button-background-color:var(--color-background);--segmented-button-text-color:var(--color-primary);--segmented-button-border-color:var(--color-primary);--segmented-button-hover-background-color:var(--color-background);--segmented-button-hover-text-color:var(--color-primary);--segmented-button-hover-border-color:var(--color-hover);--segmented-button-selected-background-color:var(--color-selected);--segmented-button-selected-text-color:var(--color-default-text);--segmented-button-selected-hover-background-color:var(--color-selected);--segmented-button-selected-hover-text-color:var(--color-default-text);--segmented-button-selected-hover-border-color:var(--color-primary);--segmented-button-border-radius:var(--radius-base)}:host([hidden]){display:none}*,*::before,*::after{box-sizing:inherit}.rux-segmented-button{display:inline-flex;height:1.6875rem;overflow:hidden;padding:0;margin:0;list-style:none;border-radius:var(--segmented-button-border-radius);background-color:var(--segmented-button-background-color)}.rux-segmented-button__segment{height:1.6875rem;width:auto;margin:0;padding:0}.rux-segmented-button__segment label{display:flex;justify-content:center;align-items:center;width:auto;height:1.5625rem;margin:0;padding:0 0.75rem;border:1px solid var(--segmented-button-border-color);color:var(--segmented-button-text-color);background-color:var(--segmented-button-background-color);font-family:var(--font-body-1-font-family);font-size:var(--font-body-1-font-size);letter-spacing:var(--font-body-1-letter-spacing);font-weight:var(--font-body-1-font-weight);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.rux-segmented-button__segment label:hover{background-color:var(--segmented-button-hover-background-color);border-color:var(--segmented-button-hover-border-color);color:var(--segmented-button-hover-text-color);outline:none}.rux-segmented-button__segment input{display:none !important}.rux-segmented-button__segment input:checked+label{background-color:var(--segmented-button-selected-background-color);color:var(--segmented-button-selected-text-color)}.rux-segmented-button__segment input:checked:hover+label{background-color:var(--segmented-button-selected-hover-background-color);color:var(--segmented-button-selected-hover-text-color);border-color:var(--segmented-button-selected-hover-border-color)}.rux-segmented-button__segment:nth-child(2):not(:last-child) label{border-right:none;border-left:none}.rux-segmented-button__segment:first-child label{border-radius:var(--segmented-button-border-radius) 0 0 var(--segmented-button-border-radius)}.rux-segmented-button__segment:last-child label{border-radius:0 var(--segmented-button-border-radius) var(--segmented-button-border-radius) 0}";

const RuxSegmentedButton = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.ruxChange = createEvent(this, "ruxchange", 7);
    /**
     * Items in this Array are the individual button segments.
     */
    this.data = [];
    /**
     * When passed in on load, this selects the first button segment with a matching label. When the selected segment changes, this property updates with the currently selected value, which reflects back to the component attribute. If no button segment label matches this string, then no segment is selected. This value takes priority over setting selected boolean property on the items in the data array.
     */
    this.selected = '';
  }
  onDataChange(newValue) {
    if (newValue) {
      const initialSelection = this.data.find((segment) => segment.selected);
      if (initialSelection) {
        this._setSelected(initialSelection.label);
      }
      else {
        if (!this.selected) {
          this._setSelected(this.data[0].label);
        }
      }
    }
  }
  connectedCallback() {
    this._handleChange = this._handleChange.bind(this);
    const initialSelection = this.data.find((segment) => segment.selected) || this.data[0];
    if (initialSelection) {
      this._setSelected(initialSelection.label);
    }
  }
  _handleChange(e) {
    const el = e.target;
    this._setSelected(el.value);
    this.ruxChange.emit(el.value);
  }
  _setSelected(label) {
    this.data.map((item) => {
      item.selected = item.label === label;
    });
    this.selected = label;
  }
  _slugify(label) {
    label = label.replace(/^\s+|\s+$/g, ''); // trim
    label = label.toLowerCase();
    label = label
      .replace(/[^a-z0-9 -]/g, '') // remove invalid chars
      .replace(/\s+/g, '-') // collapse whitespace and replace by -
      .replace(/-+/g, '-'); // collapse dashes
    return label;
  }
  _isSelected(label) {
    if (this.selected === label) {
      return true;
    }
    const selectedData = this.data.find((item) => item.label === label);
    if (selectedData && selectedData.selected) {
      return true;
    }
    return false;
  }
  render() {
    return (h("ul", { class: "rux-segmented-button" }, this.data.map((item) => (h("li", { class: "rux-segmented-button__segment" }, h("input", { type: "radio", name: "rux-group", id: this._slugify(item.label), value: item.label, checked: this._isSelected(item.label), "data-label": item.label, onChange: this._handleChange }), h("label", { htmlFor: this._slugify(item.label) }, item.label))))));
  }
  get el() { return getElement(this); }
  static get watchers() { return {
    "data": ["onDataChange"]
  }; }
};
RuxSegmentedButton.style = ruxSegmentedButtonCss;

export { RuxSegmentedButton as rux_segmented_button };
