// @astrouxds/astro-web-components/dist/esm/rux-button_2.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-01dd0ed8.js';
import { a as hasShadowDom } from './utils-0acfdbb3.js';

const ruxButtonCss = ":host{--button-text-color:var(--color-inverse-text);--button-background-color:var(--color-primary);--button-border-color:transparent;--button-hover-background-color:var(--color-hover);--button-hover-border-color:transparent;--button-active-background-color:var(--color-primary);--button-active-border-color:var(--color-primary);--button-secondary-text-color:var(--color-primary);--button-secondary-background-color:transparent;--button-secondary-border-color:var(--color-primary, rgb(0, 90, 143));--button-secondary-hover-background-color:transparent;--button-secondary-hover-border-color:var(--color-hover);display:inline-block}:host([hidden]){display:none}:host([disabled]){pointer-events:none}.rux-button{display:flex;position:relative;margin:0;padding:0 1rem;height:2.125rem;min-width:2.25rem;border-radius:var(--radius-base);color:var(--button-text-color);font-family:var(--font-body-1-font-family);font-size:var(--font-body-1-font-size);font-weight:var(--font-body-1-font-weight);letter-spacing:var(--font-body-1-letter-spacing);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;justify-content:center;align-items:center;user-select:none}.rux-button rux-icon,.rux-button ::slotted(rux-icon){height:1rem;width:1rem;margin-right:0.625rem;color:var(--button-text-color)}.rux-button:disabled{opacity:var(--disabled-opacity);cursor:not-allowed;pointer-events:none}.rux-button:focus{outline:none}.rux-button:hover:not([disabled]){cursor:pointer}.rux-button--default{border:1px solid var(--button-border-color);background-color:var(--button-background-color)}.rux-button--default:hover:not([disabled]){border-color:var(--button-hover-border-color);background-color:var(--button-hover-background-color)}.rux-button--default:active:not([disabled]){border-color:var(--button-active-border-color);background-color:var(--button-active-background-color)}.rux-button--secondary{color:var(--button-secondary-text-color);background-color:var(--button-secondary-background-color);border:1px solid var(--button-secondary-border-color)}.rux-button--secondary:hover:not([disabled]){color:var(--button-secondary-text-color);background-color:var(--button-secondary-hover-background-color);border-color:var(--button-secondary-hover-border-color)}.rux-button--secondary:active:not([disabled]){border-color:var(--button-secondary-border-color);background-color:var(--button-secondary-background-color)}.rux-button--secondary rux-icon,.rux-button--secondary ::slotted(rux-icon){color:var(--button-secondary-text-color)}.rux-button--small{height:1.625rem;padding:0 1rem;line-height:1}.rux-button--large{height:2.875rem;min-width:3rem;padding:0 1rem}.rux-button--icon-only{font-size:0;width:3rem}.rux-button--icon-only ::slotted(rux-icon),.rux-button--icon-only rux-icon{margin-left:-0.625rem;margin-right:-0.625rem}";

const RuxButton = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Hides slotted text from the button by setting rux-button--icon-only class
     */
    this.iconOnly = false;
    /**
     * Changes button style from solid to secondary by setting rux-button--secondary class
     */
    this.secondary = false;
    /**
     * Toggles disabled attribute on the button
     */
    this.disabled = false;
    /**
     * The button type. Use 'submit' to submit native form data.
     */
    this.type = 'button';
    /**
     * Creates and appends a native <button> if used within a form
     * so that it can trigger the submit event.
     *
     * Shadow DOM limitation prevents <form> from working correctly.
     * https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/button/button.tsx
     * @param ev
     */
    this.handleClick = (ev) => {
      if (hasShadowDom(this.el)) {
        const form = this.el.closest('form');
        if (form) {
          ev.preventDefault();
          const fakeButton = document.createElement('button');
          fakeButton.type = this.type;
          fakeButton.style.display = 'none';
          form.appendChild(fakeButton);
          fakeButton.click();
          fakeButton.remove();
        }
      }
    };
  }
  render() {
    const { size, iconOnly, secondary, disabled, icon } = this;
    return (h(Host, null, h("button", { type: "button", onClick: this.handleClick, class: {
        'rux-button': true,
        'rux-button--secondary': secondary,
        'rux-button--default': !secondary,
        'rux-button--small': size === 'small',
        'rux-button--large': size === 'large',
        'rux-button--icon-only': iconOnly,
      }, "aria-disabled": disabled ? 'true' : null, disabled: disabled }, icon ? (h("rux-icon", { size: "extra-small", icon: icon, color: secondary ? 'primary' : 'dark' })) : null, h("slot", null))));
  }
  get el() { return getElement(this); }
};
RuxButton.style = ruxButtonCss;

const ruxButtonGroupCss = ":host{display:inline}:host([hidden]){display:none}.rux-button-group{display:flex}.rux-button-group ::slotted(rux-button:not(:last-child)){margin-right:0.625rem}.rux-button-group--left{justify-content:flex-start}.rux-button-group--right{justify-content:flex-end}.rux-button-group--center{justify-content:center}";

const RuxButtonGroup = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * The horizontal alignment of buttons within the group
     */
    this.hAlign = 'left';
  }
  render() {
    const { hAlign } = this;
    return (h("div", { class: {
        'rux-button-group': true,
        'rux-button-group--left': hAlign === 'left',
        'rux-button-group--right': hAlign === 'right',
        'rux-button-group--center': hAlign === 'center',
      } }, h("slot", null)));
  }
};
RuxButtonGroup.style = ruxButtonGroupCss;

export { RuxButton as rux_button, RuxButtonGroup as rux_button_group };
